### PR TITLE
Fixes for PerfLog unit test in a Docker container

### PR DESCRIPTION
**Pull request by scottschurr**
Apparently when we build the RPM the process runs the unit tests as root.  That caused one of the PerfLog tests to fail because root can write to a file even if the file is write protected.

The test now verifies that removing write permissions on a file actually makes the file unwritable, and skips the test if the file is writable.

### CODE REVIEW

File: src/test/basics/PerfLog_test.cpp [hunk 1]
```diff
@@ -288,23 +289,45 @@ class PerfLog_test : public beast::unit_test::suite
                 perms::remove_perms | perms::owner_write |
                 perms::others_write | perms::group_write);
 
-            // Now construct a PerfLog.  The PerfLog should attempt to shut
-            // down the server because it can't open its file.
-            BEAST_EXPECT(parent.stopSignaled == false);
-            auto perfLog {getPerfLog (parent, WithFile::yes)};
-            BEAST_EXPECT(parent.stopSignaled == true);
+            // If the test is running as root, then the write protect may have
+            // no effect.  Make sure write protect worked before proceeding.
```
Comment (seelabs) [inline src/test/basics/PerfLog_test.cpp]: Why not directoy check if we're running as root? Something like:```
            bool const runningAsRoot = [] {
#ifndef _WIN32
                return geteuid() == 0;
#else
                return false;
#endif

File: src/test/basics/PerfLog_test.cpp [hunk 2]
```diff
@@ -288,23 +289,45 @@ class PerfLog_test : public beast::unit_test::suite
...
+                parent.doStop();
+
+                // Fix file permissions so the file can be cleaned up.
+                permissions (fullPath,
+                    perms::add_perms | perms::owner_write |
+                    perms::others_write | perms::group_write);
+            }
+            else
+            {
+                j_.fatal() << "Unable to write protect file.  Test skipped.";
```
Comment (mellery451) [inline src/test/basics/PerfLog_test.cpp]: this feel like something that belongs in a `log()` message since it's test related, but it's not a big deal either way. If you stick with the journal, should it just be a warning?

File: src/test/basics/PerfLog_test.cpp (hunk 1 above)
Comment (scottschurr) [reply src/test/basics/PerfLog_test.cpp]: I actually prefer checking whether the file is writable.

1. It's what matters.  If Perflog can write to the file then the failure won't happen.  So we're looking at what will actually break the test (or not).  The fact that this particular failure was caused by running the test as root is incidental.

2. Beyond that, the check is portable.  I don't have to worry about which operating system I'm running on.

Just my take on it.  But it's not a make-or-break thing for me.

Comment (seelabs) [reply src/test/basics/PerfLog_test.cpp]: I'm fine with that. The downside is we're cluttering up the code that can be one call with something that's more complicated, and we're really dealing with an edge case (running the unit tests as root). But again, I'm fine with keeping it as-is if you prefer it.

File: src/test/basics/PerfLog_test.cpp (hunk 2 above)
Comment (scottschurr) [reply src/test/basics/PerfLog_test.cpp]: I'm writing to the journal that `Env` provides.  It I write at `fatal` level then it shows up on the console.  I tried writing at `warn` level and the write didn't get to the console.  I felt like getting the message (that we're skipping a test and why) out to the console was a good thing.  But I could be persuaded otherwise.

Comment (mellery451) [reply src/test/basics/PerfLog_test.cpp]: the `j_` journal here is what I would call the "application log"..that's as opposed to the `log` stream that the test suite provides. This is really splitting hairs...I realize you just want to get the message out and I agree that's really what matters. Just for completeness, here's the output using `j_`:

```
ripple.basics.PerfLog
FTL:PerfLog_test Unable to create performance log directory "/tmp/perf_log_test_dir": File exists
FTL:PerfLog_test Unable to write protect file.  Test skipped.
627ms, 1 suite, 1 case, 45636 tests total, 0 failures
```

and if you instead use the test suite log stream like: `log << "Unable to write protect file.  Test skipped." << std::endl;`, then you get:

```
ripple.basics.PerfLog
FTL:PerfLog_test Unable to create performance log directory "/tmp/perf_log_test_dir": File exists
Unable to write protect file.  Test skipped.
725ms, 1 suite, 1 case, 45636 tests total, 0 failures
```

I've beaten this one to death ... and I'm perfectly fine with it either way.

File: src/test/basics/PerfLog_test.cpp (hunk 1 above)
Comment (seelabs) [reply src/test/basics/PerfLog_test.cpp]: One more note on this. It can be re-written as:
```
bool const writeProtected =
    !std::ofstream{fullPath.c_str(), std::ios::out | std::ios::app};
```
We don't need `close` as it's called by the destructor, and once that is gone there's no reason to keep the lambda.

Comment (seelabs) [reply src/test/basics/PerfLog_test.cpp]: Write protected is checked twice in this block (lines 283 through 287). How about creating one lambda to check them both. Here's a patch I have in mind: https://github.com/seelabs/rippled/commit/eb72db6b772ebe464db72995a15b1563973670f2

That's clear (`if (writeProtected(fullPath))` is very clear what's happening, and it's "8 additions and 18 deletions", which always make me feel good :-)

**Comment by scottschurr** [comment]
Thanks for the suggested change, @seelabs.  I cherry-picked it (although I'll admit I did some minor tweaking).

File: src/test/basics/PerfLog_test.cpp [hunk 3]
```diff
@@ -280,55 +280,44 @@ class PerfLog_test : public beast::unit_test::suite
             if (! BEAST_EXPECT(! ec))
                 return;
 
-            std::ofstream woFile;
-            woFile.open (fullPath.c_str(), std::ios::out | std::ios::app);
-            if (! BEAST_EXPECT(woFile))
+            auto fileWriteable = [](boost::filesystem::path const& p)
+            {
+                return std::ofstream {p.c_str(), std::ios::out | std::ios::app};
```
Comment (seelabs) [inline src/test/basics/PerfLog_test.cpp]: This lambda returns a `ofstream`. We want it to return a bool.

Comment (scottschurr) [reply src/test/basics/PerfLog_test.cpp]: Good point.  I have a fix coming.

**Comment by scottschurr** [comment]
Lambda now returns a `bool`.

**Comment by seelabs** [comment]
Still :+1: after latest commits.

**Comment by MarkusTeufelberger** [comment]
Missing this PR makes every build in https://github.com/MarkusTeufelberger/rippled-distrotest fail by the way. Looking forward to b5.

**Comment by scottschurr** [comment]
Yup.  I'm sure that's irritating.  Thanks for your patience, @MarkusTeufelberger.

**Comment by nbougalis** [comment]
Merged as 6bd0b850a06db981b9b149d82c9387db6faea7fb.
